federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+				toDistributeMax -= rdc.max
+			}
+		}
+	}
+	return toDistributeMax
+}
+
+func distributeMinReplicas(toDistributeMin int32, lists hpaLists, rdc replicaNums,
+	currentObjs map[string]pkgruntime.Object, scheduled map[string]*replicaNums) int32 {
+	for cluster := range scheduled {
+		if toDistributeMin == 0 {
+			break
+		}
+		// We have distriubted Max and thus scheduled might not be nil
+		// but probably current (what we got originally) is nil(no hpa)
+		if scheduled[cluster] == nil || currentObjs[cluster] == nil {
+			continue
+		}
+		if minReplicasAdjusteable(currentObjs[cluster]) {
+			if lists.availableMin.Len() > 0 {
+				popped, notEmpty := lists.availableMin.PopAny()
+				if notEmpty {
+					// Boundary checks have happened earlier.
+					scheduled[popped].min--
+					scheduled[cluster].min++
+					toDistributeMin--
+				}
+			}
+		}
+	}
+
+	if lists.noHpa.Len() > 0 {
+		// TODO: can this become an infinite loop?
+		for toDistributeMin > 0 {
+			for _, cluster := range lists.noHpa.UnsortedList() {
+				replicas := scheduled[cluster]
+				if replicas == nil {
+					// We did not get max here so this cluster
+					// remains without hpa
+					continue
+				}
+				var replicaNum int32 = 0
+				if toDistributeMin < rdc.min {
+					replicaNum = toDistributeMin
+				} else {
+					replicaNum = rdc.min
+				}
+				if (replicas.max - replicaNum) < replicas.min {
+					// Cannot increase the min in this cluster
+					// as it will go beyond max
+					continue
+				}
+				if lists.availableMin.Len() > 0 {
+					popped, notEmpty := lists.availableMin.PopAny()
+					if notEmpty {
+						// Boundary checks have happened earlier.
+						scheduled[popped].min--
+						replicas.min++
+						toDistributeMin--
+						continue
+					}
+				}
+				replicas.min += replicaNum
+				toDistributeMin -= replicaNum
+			}
+		}
+	} else { // we have no new clusters but if still have min replicas to distribute;
+		// just distribute all in current clusters.
+		for toDistributeMin > 0 {
+			for _, replicas := range scheduled {
+				if replicas == nil {
+					// We did not get max here so this cluster
+					// remains without hpa
+					continue
+				}
+				var replicaNum int32 = 0
+				if toDistributeMin < rdc.min {
+					replicaNum = toDistributeMin
+				} else {
+					replicaNum = rdc.min
+				}
+				if (replicas.max - replicaNum) < replicas.min {
+					// Cannot increase the min in this cluster
+					// as it will go beyond max
+					continue
+				}
+				if lists.availableMin.Len() > 0 {
+					popped, notEmpty := lists.availableMin.PopAny()
+					if notEmpty {
+						// Boundary checks have happened earlier.
+						scheduled[popped].min--
+						replicas.min++
+						toDistributeMin--
+						continue
+					}
+				}
+				replicas.min += replicaNum
+				toDistributeMin -= replicaNum
+			}
+		}
+	}
+	return toDistributeMin
+}
+
+func (a *HpaAdapter) finaliseScheduled(fedObj pkgruntime.Object, scheduled map[string]*replicaNums) (map[string]pkgruntime.Object, error) {
```

Comment:
It would be good to add a comment here explaining what this function does, and what it's parameters and return value are.